marimo/_plugins/ui/_impl/table.py

```diff
@@ -223,19 +226,26 @@ def __init__(
             self._filtered_manager = self._filtered_manager.limit(
                 TableManager.DEFAULT_ROW_LIMIT
             )
+        # get the formatted data
+        self._formatted_data = self._apply_formatting(
+            self._filtered_manager.data
+        )
+        # get the formatted manager
+        self._formatted_manager = get_table_manager(self._formatted_data)
 
         # pagination defaults to True if there are more than 10 rows
         if pagination is None:
             pagination = totalRows > 10
 
-        field_types = self._manager.get_field_types()
+        # field_types = self._manager.get_field_types()
+        field_types = self._formatted_manager.get_field_types()
 
         super().__init__(
             component_name=table._name,
             label=label,
             initial_value=[],
             args={
-                "data": self._filtered_manager.to_data(),
+                "data": self._formatted_manager.to_data(),
```

Comment:
I think we can pass the `format_mapping` directly into the table `self._filtered_manager.to_data(format_mapping),`.

And then each `TableManager` implementation can handle the formatting themselves, which should reduce the size of `_apply_formatting`

Reply:
Can we do something like `self._filtered_manager.apply_formatting(self.format_mapping).to_data()` here, and implement `apply_formatting` for each `TableManager` instead?

One benefit of this implementation is in `_convert_value`, where we can also return formatted data:

```python
def _convert_value(
    self, value: list[str]
) -> Union[List[JSONType], "pd.DataFrame", "pl.DataFrame"]:
    indices = [int(v) for v in value]
    self._selected_manager = self._filtered_manager.select_rows(indices)
    self._has_any_selection = len(indices) > 0
    return self._selected_manager.apply_formatting(self.format_mapping).data  # type: ignore[no-any-return]
```
Original `_apply_formatting`, `_format_row`, … can be removed.

Reply:
I actually don't think we want the formatted value in the conversion / result. we really only want it for displaying purposes